### PR TITLE
[music][Estuary] Add ability to play albums directly from Estuary home screen, extend context menu functionality

### PR DESCRIPTION
**Pull request by ksooo**
This is somehow a followup for #20919 

@the-black-eagle had some very cool ideas how to enhance music context menus in general and Estuary home screen music item behavior (esp. click on albums) and I offered my help to get the initial implementation into shape. The result is this PR. Thanks @the-black-eagle that you have not taken offense because I took over and restarted with a new PR (we discussed this beforehand on Slack). For me it was the best way to understand how things work.

Okay, what's in here?

* music context menu items 'Play', 'Queue', Play next' and 'Browse into' are now available not only where they appeared before this PR in the Music windows, but also in the respective Estuary Home screen items (like 'Recently played albums').

screenshot00001

* display a toast notification when something was added to the music playlist via context menu (only, if the queuing did not automatically start playback as that should be enough feedback for the user already)

screenshot00005
screenshot00002

* Add support to play complete albums directly from the Music and Song Info dialogs, making the 'script.album' external dependency in Estuary obsolete.

screenshot00006
screenshot00008

* Add an Estuary skin setting to control the default click action for music albums.

screenshot00003
screenshot00004

I tried to keep the different features in separate commits, so that review should somehow be not to difficult.

@the-black-eagle I would be thankful for comments and a runtime test/bug reports.

@enen92 might be interested in a code review.

@ronie wants to take a look at the Estuary changes I guess.

@DaveTBlake seems to be MIA, so I feel like we will not hear from him anytime soon. But in case you are listening, your feedback would be highly appreciated.

### CODE REVIEW

**Comment by ksooo** [comment]
Adding WIP label for now, as currently the new context menu items appear at some places where say should not. So, if you want to test, for now please ignore those items at 'strange' places (which will obviously not work when clicked).

**Comment by the-black-eagle** [comment]
@ksooo 

Firstly, thanks for the PR and the discussion on slack, it's always appreciated.

So, a quick test seems to show that whilst the context menu stuff works fine, playing from the home screen seems bugged :(

1 - Playback does not start with the first item.  I'm not yet sure what is going on here but it often starts with the last track rather than the first although this is not the case with albums containing many tracks.
2 - Playback from the home screen of an album does not trigger the music visualization screen when that setting is enabled, although playback from anywhere else using the context menu does.

That's just from a quick 10 minute test though.  Also, can you elaborate as to where else you are seeing the new context menu items where they shouldn't be?  Probably a visibility condition needs tweaking/adding or something.

**Comment by ksooo** [comment]
>  Also, can you elaborate as to where else you are seeing the new context menu items where they shouldn't be?

Movies/TV Shows home screen items, like top level entries, recently added movies, ... - have already found a solution and testing it locally.

> Playback does not start with the first item.

This is something related to the sort method chosen for the items to queue. There is some code I need to understand better. The sort method from the view state of WINDOW_MUSIC_NAV for the item to queue is obtained and then applied. So, the order seems to be related to what you have chosen (or the default) for the different music window views.  

> does not trigger the music visualization screen

I will take a look.

**Comment by ksooo** [comment]
> does not trigger the music visualization screen

Hmm, works here.

**Comment by the-black-eagle** [comment]
> This is something related to the sort method chosen for the items to queue. There is some code I need to understand better. The sort method from the view state of WINDOW_MUSIC_NAV for the items to queue is obtained and then applied. So, the order seems to be related to what you have chosen (or the default) for the different music window views.

Hmm, albums are sorted by artist / year (year is original release year, not that that should matter).  Tracks are sorted by track.  Just to say, the order of tracks is indeed correct, just playback starts at a (seemingly) random point, rather than the first track.

>

    does not trigger the music visualization screen

>Hmm, works here.

Really?  I tested with a fresh master with your PR on top.  Set the skin setting to "play" on the home screen.

Click on this album...

[](https://postimg.cc/RqxfYxz3)

And it starts playing (see top right) but doesn't go to the vis window.

[](https://postimg.cc/LYX7KgPT)

Settings 

[](https://postimg.cc/njg3yRpF)

**Comment by ksooo** [comment]
> does not trigger the music visualization screen

confirmed. working on it

> playback starts at a (seemingly) random point, rather than the first track.

same.

**Comment by ksooo** [comment]
@the-black-eagle, all: force-pushed. all known issues should be fixed now. please re-test.

**Comment by ksooo** [comment]
The new control 8 for MusicInfoDialog.xml should be documented in the wiki => https://kodi.wiki/view/List_of_Built_In_Controls

**Comment by the-black-eagle** [comment]
Only bug I can find now is if I navigate Music -> Files -> music directory -> artist and play an album there using the context menu, although navigating into the album shows it as sorted by track, and playback starts with track 1, the order of songs is shuffled, even though shuffle is not enabled.

For instance, playing Albert Collins' Frozen Alive, the track order becomes 5,4,6,7,1,2,3.

**Comment by ksooo** [comment]
> although navigating into the album shows it as sorted by track, and playback starts with track 1, the order of songs is shuffled

Conformed. Fix is on its way.

**Comment by ksooo** [comment]
@the-black-eagle force-pushed a fix for the sorting problem. feedback is welcome.

**Comment by the-black-eagle** [comment]
More feedback !!  If you choose 'Browse into' from the context menu on the home screen, then when you go` back`, Kodi navigates back up the nodes until finally reaching the home screen.  What should actually happen is that `back` immediately returns to the home screen.

I solved it by calling the browse into like this 
```
 const std::string target = StringUtils::Format("ActivateWindow({},{},return)", WINDOW_MUSIC_NAV,
                                                 StringUtils::Paramify(itemIn->GetPath()));
  CGUIMessage message(GUI_MSG_EXECUTE, 0, 0);
  message.SetStringParam(target);
  CServiceBroker::GetGUI()->GetWindowManager().SendMessage(message);
```
But you might know a better way to solve it.

**Comment by ksooo** [comment]
> I solved it by calling the browse into like this

 Thanks for the hint. Can indeed be coded a bit more efficient. Will do.

**Comment by ksooo** [comment]
@the-black-eagle latest force-push contains the fix for the 'back' problem.

**Review by the-black-eagle — APPROVED** [review]
Awesome, LGTM!

Thanks very much for this :)

**Comment by ksooo** [comment]
@enen92 @garbear maybe you have some thoughts on the code changes and can do a code review?
@ronie do the skin changes look okay to you? Please review.

**Comment by enen92** [comment]
I'll take a look tomorrow

**Review by ronie — APPROVED** [review]
skin changes look ok to me

**Comment by ksooo** [comment]
Heads-up: If I don't get any veto I will merge this PR mid/end of this week.

File: xbmc/music/MusicUtils.cpp
Comment (enen92) [inline xbmc/music/MusicUtils.cpp:565]: Another one that doesn't belong here, not your fault

**Review by enen92 — COMMENTED** [review]
Changes look good. Essentially there's a lot of GUI stuff on the utils class. This is not exactly your fault though, they were there already - hence it's perfectly fine if you don't want to invest more time into it.

File: xbmc/interfaces/builtins/PlayerBuiltins.cpp
Comment (ksooo) [reply xbmc/interfaces/builtins/PlayerBuiltins.cpp:535]: Not possible, playlist player methods that get called the next lines are not const.

File: xbmc/interfaces/builtins/PlayerBuiltins.cpp
Comment (ksooo) [reply xbmc/interfaces/builtins/PlayerBuiltins.cpp:579]: No, `g_application` is not gone. Some functionality is reached only via the global. `PlayMedia` is one of those.

File: xbmc/interfaces/builtins/PlayerBuiltins.cpp
Comment (ksooo) [reply xbmc/interfaces/builtins/PlayerBuiltins.cpp:785]: Which page exactly (link)?

File: xbmc/interfaces/builtins/PlayerBuiltins.cpp
Comment (enen92) [reply xbmc/interfaces/builtins/PlayerBuiltins.cpp:579]: Yeah maybe it's appPlayer that spiff encapsulated somewhere a while ago

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:226]: Not actually changed by this PR (only by the clang-format commit) - not scope of this PR.

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:240]: Not actually changed by this PR (only by the clang-format commit) - not scope of this PR.

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:275]: Not actually changed by this PR (only by the clang-format commit) - not scope of this PR.

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:309]: Not actually changed by this PR (only by the clang-format commit) - not scope of this PR.

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:345]: Same answer. Not actually changed by this PR (only by the clang-format commit) - not scope of this PR. (you should review commit by commit, omitting pure clang.format commits)

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:382]: Not actually changed by this PR (only by the clang-format commit) - not scope of this PR.

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:455]: Nope. `XFILE::CMusicDatabaseDirectory::ContainsSongs` is not const.

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:462]: Nope. `CMusicDbUrl::AppendPath`is not const.

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:484]: No way. `XFILE::CDirectory::GetDirectory`fills `items`.

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:565]: That's the question. Looking at the content of this source file I consider it "GUI".

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:19]: Looking at the content of this source file I consider it "GUI". I don't see a problem with this include.

File: xbmc/interfaces/builtins/PlayerBuiltins.cpp
Comment (ksooo) [reply xbmc/interfaces/builtins/PlayerBuiltins.cpp:579]: ... and I actually did not change that code. I just put an if clause around the existing code.

File: xbmc/interfaces/builtins/PlayerBuiltins.cpp
Comment (enen92) [reply xbmc/interfaces/builtins/PlayerBuiltins.cpp:785]: See the doc block above this (GetOperations()) function. Ends up here: https://xbmc.github.io/docs.kodi.tv/master/kodi-base/d0/d3e/page__list_of_built_in_functions.html

File: xbmc/music/MusicUtils.cpp
Comment (enen92) [reply xbmc/music/MusicUtils.cpp:565]: Looking at the atual usages of the include I'd say it's debatable. Used mainly from GUI components but also from application and infoscanners...

File: xbmc/interfaces/builtins/PlayerBuiltins.cpp
Comment (ksooo) [reply xbmc/interfaces/builtins/PlayerBuiltins.cpp:785]: Ah, I see. I will add the documentation. Thanks.

Comment (ksooo) [reply xbmc/interfaces/builtins/PlayerBuiltins.cpp:805]: done.

File: xbmc/music/MusicUtils.cpp
Comment (ksooo) [reply xbmc/music/MusicUtils.cpp:565]: Call hierarchy shows that `ShowToastNotification`is purely GUI and not revealed in the header.

Screenshot 2022-10-26 at 08 51 09

I agree that MusicUtils.h could be refactored into a GUI and a non-GUI part, but this is for another day (PR).

**Comment by ksooo** [comment]
Final rebase done. I will merge tomorrow if nobody objects until then.

**Comment by ksooo** [comment]
jenkins build this please

**Pull request merged**
